usr/lib/cca_stdll/cca_specific.c

```diff
@@ -2974,6 +2974,304 @@ static CK_BBOOL ccatok_pkey_option_disabled(STDLL_TokData_t *tokdata)
     return CK_FALSE;
 }
 
+/**
+ * Return true, if the given key obj has a valid protected key, i.e. its
+ * verification pattern matches the one of the current master key.
+ */
+static CK_BBOOL ccatok_pkey_is_valid(STDLL_TokData_t *tokdata, OBJECT *key_obj)
+{
+    struct cca_private_data *cca_data = tokdata->private_data;
+    CK_ATTRIBUTE *pkey_attr = NULL;
+    int vp_offset;
+
+    if (template_attribute_get_non_empty(key_obj->template, CKA_IBM_OPAQUE_PKEY,
+                                         &pkey_attr) == CKR_OK) {
+        if (pkey_attr->ulValueLen >= AES_KEY_SIZE_128 + PKEY_MK_VP_LENGTH) {
+            vp_offset = pkey_attr->ulValueLen - PKEY_MK_VP_LENGTH;
+            if (memcmp((CK_BYTE *)pkey_attr->pValue + vp_offset,
+                       &cca_data->pkey_mk_vp,
+                       PKEY_MK_VP_LENGTH) == 0) {
+                return CK_TRUE;
+            }
+        }
+    }
+
+    return CK_FALSE;
+}
+
+/**
+ * Create a new protected key for the given key obj and update attribute
+ * CKA_IBM_OPAQUE with the new pkey.
+ */
+static CK_RV ccatok_pkey_update(STDLL_TokData_t *tokdata, OBJECT *key_obj)
+{
+    struct cca_private_data *cca_data = tokdata->private_data;
+    CK_ATTRIBUTE *skey_attr = NULL;
+    CK_ATTRIBUTE *pkey_attr = NULL;
+    CK_RV ret;
+    int vp_offset;
+
+    /* Get secure key from obj */
+    if (template_attribute_get_non_empty(key_obj->template, CKA_IBM_OPAQUE,
+                                         &skey_attr) != CKR_OK) {
+        TRACE_ERROR("This key has no blob: should not occur!\n");
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Transform the secure key into a protected key */
+    ret = ccatok_pkey_skey2pkey(tokdata, skey_attr, &pkey_attr);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("protected key creation failed with rc=0x%lx\n",ret);
```

Comment:
Here you must take care about the case when it is failing because a concurrent MK change is active, and none of the APQNs have the desired MK to convert the blobs to protected keys. In this case the ioctl will return ENODEV I guess, but this will lead to CKR_FUNCTION_FAILED here. 
In such a case (quite likely during MK changes), you should NOT just fail, but fall back to secure key usage.

Reply:
If we think into the future about AES-XTS support, that has no secure key fallback, we might want to better handle the MK change case, and add code to retry the conversion with the reenciphered blobs. But lets keep that in mind for now, and maybe add it on top of this patch queue.